pkg/autodiscovery/listeners/kube_services.go

```diff
@@ -225,14 +265,232 @@ func (l *KubeServiceListener) removeService(ksvc *v1.Service) {
 	if ok {
 		l.m.Lock()
 		delete(l.services, ksvc.UID)
+		delete(l.endpointsAnnotated, fmt.Sprintf("%s/%s", ksvc.Namespace, ksvc.Name))
 		l.m.Unlock()
 
 		l.delService <- svc
+
 	} else {
 		log.Debugf("Entity %s not found, not removing", ksvc.UID)
 	}
 }
 
+func (l *KubeServiceListener) addedEndpt(obj interface{}) {
+	castedObj, ok := obj.(*v1.Endpoints)
+	if !ok {
+		log.Errorf("Expected a Endpoints type, got: %v", obj)
+		return
+	}
+	l.createEndpoint(castedObj, false)
+}
+
+func (l *KubeServiceListener) deletedEndpt(obj interface{}) {
+	castedObj, ok := obj.(*v1.Endpoints)
+	if !ok {
+		log.Errorf("Expected a Endpoint type, got: %v", obj)
+		return
+	}
+	l.removeEndpoint(castedObj)
+}
+
+func (l *KubeServiceListener) updatedEndpt(old, obj interface{}) {
+	// Cast the updated object or return on failure
+	castedObj, ok := obj.(*v1.Endpoints)
+	if !ok {
+		log.Errorf("Expected a Endpoints type, got: %v", obj)
+		return
+	}
+	// Cast the old object, consider it an add on cast failure
+	castedOld, ok := old.(*v1.Endpoints)
+	if !ok {
+		log.Errorf("Expected a Endpoints type, got: %v", old)
+		l.createEndpoint(castedObj, false)
+		return
+	}
+	if endpointsDiffer(castedObj, castedOld) {
+		l.removeEndpoint(castedObj)
+		l.createEndpoint(castedObj, false)
+	}
+}
+
+// endpointsDiffer compares two endpoints to only go forward
+// when relevant fields are changed. This logic must be
+// updated if more fields are used.
+func endpointsDiffer(first, second *v1.Endpoints) bool {
+	// Quick exit if resversion did not change
+	if first.ResourceVersion == second.ResourceVersion {
+		return false
+	}
+
+	// Subsets
+	if len(first.Subsets) != len(second.Subsets) {
+		return true
+	}
+
+	// Addresses and Ports
+	for i := range first.Subsets {
+		if len(first.Subsets[i].Addresses) != len(second.Subsets[i].Addresses) {
+			return true
+		}
+		if len(first.Subsets[i].Ports) != len(second.Subsets[i].Ports) {
+			return true
+		}
+		// Addresses
+		for j := range first.Subsets[i].Addresses {
+			if first.Subsets[i].Addresses[j].IP != second.Subsets[i].Addresses[j].IP {
+				return true
+			}
+			if first.Subsets[i].Addresses[j].Hostname != second.Subsets[i].Addresses[j].Hostname {
```

Comment:
I'm curious how this code would benchmark against calling `reflect.DeepEqual` on the `Subsets` objects. Could you test it?

Reply:
good reflex, thanks 👍 
```
goos: linux
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/autodiscovery/listeners
BenchmarkEndpointsDiffer/Remove_port-4         	300000000	         5.90 ns/op
BenchmarkEndpointsDiffer/Same_resversion-4     	500000000	         3.21 ns/op
BenchmarkEndpointsDiffer/Change_resversion,_same_subsets-4         	50000000	        25.9 ns/op
BenchmarkEndpointsDiffer/Change_IP-4                               	100000000	        15.5 ns/op
BenchmarkEndpointsDiffer/Change_Hostname-4                         	100000000	        15.6 ns/op
BenchmarkEndpointsDiffer/Change_port_number-4                      	50000000	        23.4 ns/op
BenchmarkEndpointsDiffer/Remove_IP-4                               	300000000	         5.42 ns/op
BenchmarkEndpointsDifferDeepEqual/Change_Hostname-4                	 2000000	       796 ns/op
BenchmarkEndpointsDifferDeepEqual/Change_port_number-4             	 1000000	      1841 ns/op
BenchmarkEndpointsDifferDeepEqual/Remove_IP-4                      	10000000	       209 ns/op
BenchmarkEndpointsDifferDeepEqual/Remove_port-4                    	 1000000	      1266 ns/op
BenchmarkEndpointsDifferDeepEqual/Same_resversion-4                	500000000	         3.44 ns/op
BenchmarkEndpointsDifferDeepEqual/Change_resversion,_same_subsets-4         	 1000000	      1956 ns/op
BenchmarkEndpointsDifferDeepEqual/Change_IP-4                               	 2000000	       807 ns/op
```